### DISCUSSION TITLE
Question about JSON RPC options

### DISCUSSION THREAD

**Discussion by vetcher** [go-kit/kit#1140] [Q&A]
Is there any way to pass individual options to each endpoint from `EndpointCodecMap`? I want to pass `opentracing.HTTPToContext`, that requires operationName.  

Also, I think, [jsonrpc example](https://github.com/go-kit/kit/blob/master/examples/addsvc/pkg/addtransport/jsonrpc.go) may be improved, by using `HTTPToContext` and `ContextToHTTP` functions from opentracing, like it done in [grpc](https://github.com/go-kit/kit/blob/master/examples/addsvc/pkg/addtransport/grpc.go#L57) and [http](https://github.com/go-kit/kit/blob/master/examples/addsvc/pkg/addtransport/http.go#L59) implementations. @rossmcf

**Comment by rossmcf** [comment]
Hi @Vetcher,

Good question. No, there's no way to handle that per-endpoint at the moment, so I think we may need to build that in. At a glance, this could be done by having before/after `RequestFunc`s as fields of the `EndpointCodecMap`, but there may well be a more elegant way.

I'm not very familiar with the requirements for tracing (hence this omission from the transport), but would love to see it added. Would you be up for submitting a PR?

**Comment by basvanbeek** [comment]
so this is also a question on where what fits...

the question we need to ask ourselves is... do we want serverbefore middleware to be able to block / handle requests before the endpoint mapping is done or do we want to do this after mapping (so we add serverbefore payloads per JSON-RPC method.

An alternative solution is to keep as is but inject the selected JSON-RPC method as value into context. With a minor modification to TraceServer we can alter the span's operation name using the value found in context. By doing this we wouldn't break existing APIs.

**Comment by vetcher** [comment]
@basvanbeek there is problem of alternative solution, that you proposed. ServerBefore functions executes before extracting name of rpc method from input json.

I'm thinking about adding to `EndpointCodec` slice of serverbefore functions, but as @rossmcf sad, it feels like not the best solution.

**Comment by basvanbeek** [comment]
that can be fixed on the server end by doing the first JSON-RPC request decode before handling the before functions...
